Wallets Service & DiscoverAsset Service updates

gemdev111 opened the pull request:
Removed combine from DiscoverAssetsService & WalletsService
WalletsService migrated to structured Concurrency (safe parallelism)
BalanceService migrated to structured Concurrency
Removed unused imports & minor code cleanup, confirming Sendable


closes #409 
closes #411

gemdev111 committed:
Removed combine from DiscoverAssetsService & WalletsService
WalletsService migrated to structured Concurrency (safe parallelism)
BalanceService migrated to structured Concurrency
Removed unused imports & minor code clenup, confirming Sendable

gemdev111 committed:
Update core

gemdev111 committed:
Added missed Sendalbe
Added missed weak self for closures, to avoid strong reference cycles

gemcoder21 commented on Services/DiscoverAssetsService/Sources/Types/AssetUpdate.swift:
```diff
@@ -0,0 +1,9 @@
+import Primitives
+
+public struct AssetUpdate: Sendable {
+    public let wallet: Wallet
```
```suggestion
    public let walletId: WalletId
```

gemcoder21 commented on Services/WalletsService/Sources/WalletsService.swift:
```diff
@@ -155,67 +132,88 @@ public final class WalletsService: @unchecked Sendable {
+            wallet: wallet,
+            fromTimestamp: prefs.assetsTimestamp
+        )
+        prefs.assetsTimestamp = newTimestamp
```
```suggestion
        preferences.assetsTimestamp = newTimestamp
```

gemcoder21 commented on Services/BalanceService/Sources/BalanceService.swift:
```diff
@@ -9,7 +9,6 @@ import ChainService
 public final class BalanceService: BalancerUpdater, Sendable {
```
should to struct

gemcoder21 commented on Services/BalanceService/Sources/BalanceService.swift:
```diff
@@ -69,42 +68,41 @@ extension BalanceService {
-            guard !ids.isEmpty else {
-                continue
-            }
-            // coin balance
```
add comments back

gemcoder21 commented on Services/WalletsService/Sources/WalletsService.swift:
```diff
@@ -155,67 +132,88 @@ public final class WalletsService: @unchecked Sendable {
-                await discoverAssetService.updateCoins(wallet: wallet)
-                preferences.completeInitialLoadAssets = true
-            }
+    private func processCoinDiscovery(for wallet: Wallet, prefs: WalletPreferences) async throws {
```
add mini services with protocols for each process

gemcoder21 commented on Services/WalletsService/Sources/WalletsService.swift:
```diff
@@ -155,67 +132,88 @@ public final class WalletsService: @unchecked Sendable {
 
-        //FIXME: temp solution to wait for 1 second (need to wait until subscriptions updated)
-        //otherwise it would return no assets
-        try await Task.sleep(nanoseconds: UInt64(1 * Double(NSEC_PER_SEC)))
```
add `try await Task.sleep(nanoseconds: UInt64(1 * Double(NSEC_PER_SEC)))` back

gemcoder21 commented on Services/DiscoverAssetsService/Sources/DiscoverAssetsService.swift:
```diff
@@ -30,41 +21,74 @@ public struct DiscoverAssetsService: @unchecked Sendable {
             deviceId: deviceId,
             walletIndex: wallet.index.asInt,
             fromTimestamp: fromTimestamp
         )
```
add `.compactMap here { if let address = try? wallet.account(for: assetId.chain).address,
                   assetId.type == .token  }`

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
BalanceService as struct
DiscoverAssetsService as struct
Removed weakself
Simplify code DiscoverAssetsService
Updates AssetUpdate model from wallet => walletId

gemdev111 committed:
prefs naming updates

gemdev111 committed:
WalletsService to struct, remvoed exra weakselfs
Added sleep() for processTokenDiscovery to make sure subscriptions updated

gemdev111 committed:
Added protocols for WalletsService
Codebase cleanup & updates
